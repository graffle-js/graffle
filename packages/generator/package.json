{
  "name": "@graffle/generator",
  "version": "0.1.0",
  "description": "Code generator library for Graffle GraphQL clients",
  "type": "module",
  "exports": {
    ".": "./build/index.js",
    "./helpers/standard-scalar-types": "./build/helpers/standard-scalar-types.js"
  },
  "imports": {
    "#~/*": "./build/*"
  },
  "files": ["build", "src"],
  "scripts": {
    "build": "tsc --build tsconfig.build.json",
    "build:emit": "tsc --build tsconfig.build.json --noCheck",
    "build:clean": "tsc --build tsconfig.build.json --clean && rm -rf build",
    "dev": "tsc --build tsconfig.build.json --watch --noCheck",
    "check": "pnpm run '/check:.*/'",
    "check:types": "tsc --noEmit",
    "test": "vitest",
    "test:unit": "vitest --run"
  },
  "dependencies": {
    "@graffle/core": "workspace:*",
    "@graffle/graphql": "workspace:*",
    "@graffle/extension-introspection": "workspace:*",
    "@wollybeard/kit": "^0.98.0",
    "ts-morph": "^27.0.2"
  },
  "peerDependencies": {
    "@dprint/formatter": "^0.4.0",
    "@dprint/typescript": "^0.93.0",
    "prettier": "^3.3.3"
  },
  "peerDependenciesMeta": {
    "prettier": { "optional": true },
    "@dprint/formatter": { "optional": true },
    "@dprint/typescript": { "optional": true }
  },
  "devDependencies": {
    "@test/fixtures": "workspace:*"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graffle-js/graffle.git",
    "directory": "packages/generator"
  },
  "keywords": ["graphql", "generator", "codegen"],
  "author": {
    "name": "Jason Kuhrt",
    "email": "jasonkuhrt@me.com",
    "url": "https://kuhrt.me"
  },
  "license": "MIT"
}
