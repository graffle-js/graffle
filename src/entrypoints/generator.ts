export { generateFiles as generate } from '../layers/4_generator/files.js'
