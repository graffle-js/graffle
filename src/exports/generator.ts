export * from './__Generator.js'
