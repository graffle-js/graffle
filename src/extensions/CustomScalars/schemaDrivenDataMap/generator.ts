import type { Config } from '../../../layers/4_generator/config/config.js'
import { ModuleGeneratorScalar } from '../../../layers/4_generator/generators/Scalar.js'
import { createModuleGenerator } from '../../../layers/4_generator/helpers/moduleGenerator.js'
import { createCodeGenerator } from '../../../layers/4_generator/helpers/moduleGeneratorRunner.js'
import { title1 } from '../../../layers/4_generator/helpers/render.js'
import { Code } from '../../../lib/Code.js'
import { Grafaid } from '../../../lib/grafaid/__.js'
import { entries } from '../../../lib/prelude.js'
import { nullabilityFlags, propertyNames } from './types.js'

const identifiers = {
  $Scalar: `$Scalar`,
}

type ReferenceAssignments = string[]

export const ModuleGeneratorSchemaDrivenDataMap = createModuleGenerator(
  `SchemaDrivenDataMap`,
  ({ config, code }) => {
    code(`
      import * as ${identifiers.$Scalar} from './${ModuleGeneratorScalar.name}.js'
      import type * as $Utilities from '${config.paths.imports.grafflePackage.utilitiesForGenerated}'
    `)

    const kinds = getKinds(config)
    const kindsList = entries(kinds)

    const referenceAssignments: ReferenceAssignments = []

    for (const [kindName, nodes] of kindsList) {
      code(title1(kindName))
      code()
      if (nodes.length === 0) {
        code(`// None of your ${kindName}s have custom scalars.`)
      }
      for (const type of nodes) {
        const codeGenerator = kindRenders[kindName] as any
        code(codeGenerator({ config, type: type as any, referenceAssignments }))
        code()
      }
      code()
    }

    code(title1(`Reference Assignments`, `(avoids circular assignment issues)`))
    code()
    if (referenceAssignments.length === 0) {
      code(`// None of your types have references to other non-scalar/enum types.`)
    }
    for (const referenceAssignment of referenceAssignments) {
      code(referenceAssignment)
    }
    code()

    code(title1(`Index`))
    code()
    code(`const $schemaDrivenDataMap: $Utilities.SchemaDrivenDataMap =`)
    code(Code.termObject({
      roots: Code.directiveTermObject({
        $literal: kinds.GraphQLRootType.map(type => type.name + `,`).join(`\n`),
      }),
      directives: `{}`,
      types: Code.directiveTermObject({
        $literal: [
          ...kindsList.map(([, _]) => _).flat().map((_) => _.name),
          // We also include the custom scalars here to facilitate encoding. Encoding has names of variables and
          // that need to be looked up to determine which are/have custom scalars.
          // ...config.schema.typeMapByKind.GraphQLScalarTypeCustom.map(_ => {
          //   return `${_.name}: ${identifiers.$CustomScalars}.${_.name},`
          // }),
        ].join(`,\n`),
      }),
    }))
    code()
    code(`export { $schemaDrivenDataMap as schemaDrivenDataMap }`)
  },
)

//
//
//
// Helpers
// -------
//
//
//

/**
 * Get minimal kind set for SDDM
 *
 * If feature operationVariables is enabled then we need to emit all paths to inputs.
 * If feature customScalars is enabled then we need to emit all paths to customs scalar inputs AND outputs.
 * If both are enabled the merged requirement is: all paths to inputs AND custom scalar outputs.
 */
const getKinds = (config: Config) => {
  const { schema: { kindMap } } = config
  const condition = typeCondition(config)
  return {
    // When "variables" enabled, we need to know all named types to be able to write them out.
    GraphQLScalarTypeStandard: kindMap.GraphQLScalarTypeStandard.filter(() =>
      config.runtimeFeatures.operationVariables
    ),
    GraphQLScalarTypeCustom: kindMap.GraphQLScalarTypeCustom.filter(() => config.runtimeFeatures.customScalars),
    GraphQLEnumType: kindMap.GraphQLEnumType.filter(() => config.runtimeFeatures.operationVariables),
    GraphQLInputObjectType: kindMap.GraphQLInputObjectType.filter(condition),
    GraphQLObjectType: kindMap.GraphQLObjectType.filter(condition),
    GraphQLInterfaceType: kindMap.GraphQLInterfaceType.filter(condition),
    GraphQLUnionType: kindMap.GraphQLUnionType.filter(condition),
    GraphQLRootType: kindMap.GraphQLRootType.filter(condition),
  }
}

const typeCondition = (config: Config) => {
  if (config.runtimeFeatures.customScalars) {
    if (config.runtimeFeatures.operationVariables) {
      const isHasInputOrOutputCustomScalar = () => true // todo
      return isHasInputOrOutputCustomScalar
    }
    return Grafaid.Schema.CustomScalars.isHasCustomScalars
  }

  if (config.runtimeFeatures.operationVariables) {
    const isHasInput = () => true // todo
    return isHasInput
  }

  return falseFilter
}

const falseFilter = () => false
const trueFilter = () => true

const inputTypeCondition = (config: Config) => {
  if (config.runtimeFeatures.operationVariables) return trueFilter

  if (config.runtimeFeatures.customScalars) {
    return Grafaid.Schema.CustomScalars.isHasCustomScalars
  }

  return falseFilter
}

//
//
//
// Code Generators
// ---------------
//
//
//
const ScalarType = createCodeGenerator<
  { type: Grafaid.Schema.ScalarType }
>(
  ({ code, type }) => {
    code(Code.termConst(type.name, `${identifiers.$Scalar}.${type.name}`))
  },
)

const UnionType = createCodeGenerator<
  { type: Grafaid.Schema.UnionType; referenceAssignments: ReferenceAssignments }
>(
  ({ code, type }) => {
    // This takes advantage of the fact that in GraphQL, in a union type, all members that happen
    // to have fields of the same name, those fields MUST be the same type.
    // See:
    // - https://github.com/graphql/graphql-js/issues/1361
    // - https://stackoverflow.com/questions/44170603/graphql-using-same-field-names-in-different-types-within-union
    //
    // So what we do is inline all the custom scalar paths of all union members knowing
    // that they could never conflict.
    code(Code.termConstTyped(
      type.name,
      `$Utilities.SchemaDrivenDataMap.OutputObject`,
      Code.termObject({
        [propertyNames.f]: Code.directiveTermObject({
          $spread: type.getTypes().filter(Grafaid.Schema.CustomScalars.isHasCustomScalars).map(memberType =>
            memberType.name + `.${propertyNames.f}`
          ),
        }),
      }),
    ))
  },
)

const InterfaceType = createCodeGenerator<
  { type: Grafaid.Schema.InterfaceType; referenceAssignments: ReferenceAssignments }
>(
  ({ code, type, config }) => {
    const implementorTypes = Grafaid.Schema.KindMap.getInterfaceImplementors(config.schema.kindMap, type)
    code(Code.termConstTyped(
      type.name,
      `$Utilities.SchemaDrivenDataMap.OutputObject`,
      Code.termObject({
        [propertyNames.f]: Code.directiveTermObject({
          $spread: implementorTypes.filter(Grafaid.Schema.CustomScalars.isHasCustomScalars).map(memberType =>
            memberType.name + `.${propertyNames.f}`
          ),
        }),
      }),
    ))
  },
)

const ObjectType = createCodeGenerator<
  { type: Grafaid.Schema.ObjectType; referenceAssignments: ReferenceAssignments }
>(
  ({ config, code, type, referenceAssignments }) => {
    const o: Code.TermObject = {}

    config.extensions.forEach(_ => {
      _.schemaDrivenDataMap?.onObjectType?.({
        config,
        sddmNode: o,
        graphqlType: type,
      })
    })

    // Fields of this object.
    // ---------------------
    const of: Code.TermObject = {}
    o[propertyNames.f] = of

    const condition = typeCondition(config)

    const outputFields = Object.values(type.getFields()).filter(condition)
    for (const outputField of outputFields) {
      const outputFieldNamedType = Grafaid.Schema.getNamedType(outputField.type)
      const sddmNodeOutputField: Code.DirectiveTermObjectLike<Code.TermObject> = {
        $fields: {},
      }
      of[outputField.name] = sddmNodeOutputField

      // Field Arguments
      const inputCondition = inputTypeCondition(config)
      const args = outputField.args.filter(inputCondition)
      if (args.length > 0) {
        const ofItemAs: Code.TermObject = {}
        sddmNodeOutputField.$fields[propertyNames.a] = ofItemAs

        for (const arg of args) {
          const ofItemA: Code.TermObject = {}
          ofItemAs[arg.name] = ofItemA

          const argType = Grafaid.Schema.getNamedType(arg.type)
          // dprint-ignore
          if (
             (config.runtimeFeatures.customScalars && Grafaid.Schema.isScalarTypeAndCustom(argType)) ||
             // For variables, we need to know the variable type to write it out, so we always need the named type.
             (config.runtimeFeatures.operationVariables)
          ) {
              ofItemA[propertyNames.nt] = argType.name
              // For variables, we need to know the variable type to write it out, so we always need the inline type.
              if (config.runtimeFeatures.operationVariables) {
                ofItemA[propertyNames.it] = inlineType(arg.type)
              }
            }
        }
      }

      config.extensions.forEach(_ => {
        _.schemaDrivenDataMap?.onOutputField?.({
          config,
          sddmNode: sddmNodeOutputField,
          graphqlType: outputField,
        })
      })

      if (condition(outputFieldNamedType)) {
        if (Grafaid.Schema.isScalarTypeAndCustom(outputFieldNamedType)) {
          if (config.runtimeFeatures.customScalars) {
            sddmNodeOutputField.$fields[propertyNames.nt] = outputFieldNamedType.name
          }
        } else if (
          Grafaid.Schema.isUnionType(outputFieldNamedType) || Grafaid.Schema.isObjectType(outputFieldNamedType)
          || Grafaid.Schema.isInterfaceType(outputFieldNamedType)
        ) {
          referenceAssignments.push(`${type.name}.f[\`${outputField.name}\`]!.nt = ${outputFieldNamedType.name}`)
          // dprint-ignore
          sddmNodeOutputField.$literal = `// ${Code.termField(propertyNames.nt, outputFieldNamedType.name)} <-- Assigned later to avoid potential circular dependency.`
          // // todo make kitchen sink schema have a pattern where this code path will be traversed.
          // // We just need to have arguments on a field on a nested object.
          // // Nested objects that in turn have custom scalar arguments
          // if (Schema.isObjectType(fieldType) && Schema.isHasCustomScalars(fieldType)) {
          //   code(Code.termField(field.name, fieldType.name))
          // }
        }
      }
    }

    code(Code.termConstTyped(type.name, `$Utilities.SchemaDrivenDataMap.OutputObject`, Code.termObject(o)))
  },
)

const EnumType = createCodeGenerator<
  { type: Grafaid.Schema.EnumType; referenceAssignments: ReferenceAssignments }
>(
  ({ code, type }) => {
    code(Code.termConstTyped(
      type.name,
      `$Utilities.SchemaDrivenDataMap.Enum`,
      Code.termObject({
        [propertyNames.k]: Code.string(`enum`),
        [propertyNames.n]: Code.string(type.name),
      }),
    ))
  },
)

const InputObjectType = createCodeGenerator<
  { type: Grafaid.Schema.ObjectType; referenceAssignments: ReferenceAssignments }
>(
  ({ config, code, type, referenceAssignments }) => {
    const o: Code.TermObject = {}

    const inputFields = Object.values(type.getFields())

    if (config.runtimeFeatures.operationVariables) {
      o[propertyNames.n] = Code.string(type.name)
      const customScalarFields = inputFields
        .filter(Grafaid.Schema.CustomScalars.isHasCustomScalarInputs)
        .map(inputField => inputField.name)
        .map(Code.string)
      if (customScalarFields.length) {
        o[propertyNames.fcs] = Code.termList(customScalarFields)
      }
    }
    const f: Code.TermObjectOf<Code.DirectiveTermObjectLike<Code.TermObject>> = {}
    o[propertyNames.f] = f

    for (const inputField of inputFields) {
      const inputFieldType = Grafaid.Schema.getNamedType(inputField.type)

      // dprint-ignore
      const isPresent =
        config.runtimeFeatures.operationVariables ||
        (config.runtimeFeatures.customScalars &&
          (Grafaid.Schema.isScalarTypeAndCustom(inputFieldType) ||
          (Grafaid.Schema.isInputObjectType(inputFieldType) && Grafaid.Schema.CustomScalars.isHasCustomScalarInputs(inputFieldType))))

      if (!isPresent) continue

      f[inputField.name] = {
        $fields: {},
      }

      if (Grafaid.Schema.isScalarTypeAndCustom(inputFieldType)) {
        f[inputField.name]!.$fields[propertyNames.nt] = inputFieldType.name
      } else if (
        Grafaid.Schema.isInputObjectType(inputFieldType)
        && Grafaid.Schema.CustomScalars.isHasCustomScalarInputs(inputFieldType)
      ) {
        referenceAssignments.push(
          `${type.name}.${propertyNames.f}![\`${inputField.name}\`]!.${propertyNames.nt} = ${inputFieldType.name}`,
        )
        f[inputField.name]!.$literal = `// ${
          Code.termField(propertyNames.nt, inputFieldType.name)
        } <-- Assigned later to avoid potential circular dependency.`
      }
    }

    code(Code.termConstTyped(type.name, `$Utilities.SchemaDrivenDataMap.InputObject`, Code.termObject(o)))
  },
)

const kindRenders = {
  GraphQLScalarTypeStandard: ScalarType,
  GraphQLScalarTypeCustom: ScalarType,
  GraphQLEnumType: EnumType,
  GraphQLUnionType: UnionType,
  GraphQLInterfaceType: InterfaceType,
  GraphQLInputObjectType: InputObjectType,
  GraphQLObjectType: ObjectType,
  GraphQLRootType: ObjectType,
}

const inlineType = (type: Grafaid.Schema.InputTypes): string => {
  const [ofType, nonNull] = Grafaid.Schema.isNonNullType(type)
    ? [type.ofType, true]
    : [type, false]

  const nullFlag = nonNull
    ? nullabilityFlags.nonNull
    : nullabilityFlags.nullable

  const rest = Grafaid.Schema.isListType(ofType)
    ? inlineType(ofType.ofType)
    : ``

  return `[${nullFlag.toString()}, ${rest}]`
}
