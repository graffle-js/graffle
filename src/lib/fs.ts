import fs from 'node:fs/promises'
import { extname, isAbsolute, join } from 'node:path'
import { errorFromMaybeError } from './prelude.js'

export const statMaybeExists = async (path: string) => {
  return await fs.stat(path).catch((_: unknown) => {
    const error = errorFromMaybeError(_)
    return `code` in error && typeof error.code === `string` && error.code === `ENOENT`
      ? null
      : Promise.reject(error)
  })
}

export const fileExists = async (path: string) => {
  return Boolean(await statMaybeExists(path))
}

export const isPathToADirectory = async (path: string) => {
  const stat = await fs.stat(path)
  return stat.isDirectory()
}

export const toAbsolutePath = (cwd: string, maybeAbsolutePath: string) =>
  isAbsolute(maybeAbsolutePath) ? maybeAbsolutePath : join(cwd, maybeAbsolutePath)

const isFileLikePath = (path: string) => {
  return Boolean(extname(path))
}

export const toFilePath = (fileName: string, path: string) => {
  if (isFileLikePath(path)) {
    return path
  } else {
    return join(path, fileName)
  }
}

// export const toFileChecked = async (fileName: string, path: string) => {
//   if (isFileLikePath(path)) {
//     return path
//   } else {
//     return join(path, fileName)
//   }
// }
