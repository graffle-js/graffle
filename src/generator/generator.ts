import type {
  GraphQLArgument,
  GraphQLEnumValue,
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
} from 'graphql'
import { GraphQLNonNull, isEnumType, isListType, isNamedType } from 'graphql'
import { buildSchema } from 'graphql'
import _ from 'json-bigint'
import fs from 'node:fs/promises'
import { Code } from '../lib/Code.js'
import type {
  AnyClass,
  AnyField,
  AnyNamedClassName,
  ClassToName,
  Describable,
  NameToClassNamedType,
  TypeMapByKind,
} from '../lib/graphql.js'
import {
  getNodeDisplayName,
  getTypeMapByKind,
  isDeprecatableNode,
  isGraphQLOutputField,
  type NameToClass,
} from '../lib/graphql.js'
import { entries, values } from '../lib/prelude.js'

const namespaceNames = {
  GraphQLEnumType: `Enum`,
  GraphQLInputObjectType: `InputObject`,
  GraphQLInterfaceType: `Interface`,
  GraphQLObjectType: `Object`,
  GraphQLScalarType: `Scalar`,
  GraphQLUnionType: `Union`,
} satisfies Record<AnyNamedClassName, string>

type AnyGraphQLFieldsType =
  | GraphQLObjectType
  | GraphQLInterfaceType
  | GraphQLInputObjectType

const defineReferenceRenderers = <
  $Renderers extends { [ClassName in keyof NameToClass]: any },
>(
  renderers: {
    [ClassName in keyof $Renderers]: (
      config: Config,
      node: ClassName extends keyof NameToClass ? InstanceType<NameToClass[ClassName]>
        : never,
    ) => string
  },
) => renderers

const defineConcreteRenderers = <
  $Renderers extends { [ClassName in keyof NameToClassNamedType]: any },
>(
  renderers: {
    [ClassName in keyof $Renderers]: (
      config: Config,
      node: ClassName extends keyof NameToClassNamedType ? InstanceType<NameToClassNamedType[ClassName]>
        : never,
    ) => string
  },
): {
  [ClassName in keyof $Renderers]: (
    node: ClassName extends keyof NameToClass ? InstanceType<NameToClass[ClassName]> | null | undefined
      : never,
  ) => string
} => {
  return Object.fromEntries(
    Object.entries(renderers).map(([key, renderer]) => {
      return [
        key,
        (config: Config, node: any) => {
          if (!node) return ``
          return renderer(config, node) // eslint-disable-line
        },
      ]
    }),
  ) as any
}

const dispatchToReferenceRenderer = (config: Config, node: AnyClass): string =>
  getReferenceRenderer(node)(config, node as any)

const getReferenceRenderer = <N extends AnyClass>(node: N): (typeof referenceRenderers)[ClassToName<N>] => {
  // @ts-expect-error lookup
  const renderer = referenceRenderers[node.constructor.name] // eslint-disable-line
  if (!renderer) {
    throw new Error(`No renderer found for class: ${node.constructor.name}`)
  }
  return renderer
}

const referenceRenderers = defineReferenceRenderers({
  GraphQLNonNull: (config, node) => dispatchToReferenceRenderer(config, node.ofType),
  GraphQLEnumType: (_, node) => Code.propertyAccess(namespaceNames.GraphQLEnumType, node.name),
  GraphQLInputObjectType: (_, node) => Code.propertyAccess(namespaceNames.GraphQLInputObjectType, node.name),
  GraphQLInterfaceType: (_, node) => Code.propertyAccess(namespaceNames.GraphQLInterfaceType, node.name),
  GraphQLObjectType: (_, node) => Code.propertyAccess(namespaceNames.GraphQLObjectType, node.name),
  GraphQLUnionType: (_, node) => Code.propertyAccess(namespaceNames.GraphQLUnionType, node.name),
  GraphQLList: (config, node) => `_.List<${(buildType(config, node.ofType))}>`,
  GraphQLScalarType: (_, node) => `_.Scalar.${node.name}`,
})

const dispatchToConcreteRenderer = (
  config: Config,
  node: GraphQLNamedType,
): string => {
  // @ts-expect-error lookup
  const renderer = concreteRenderers[node.constructor.name] // eslint-disable-line
  if (!renderer) {
    throw new Error(`No renderer found for class: ${node.constructor.name}`)
  }
  return renderer(config, node) // eslint-disable-line
}

const concreteRenderers = defineConcreteRenderers({
  GraphQLEnumType: (config, node) =>
    Code.TSDoc(
      getDocumentation(config, node),
      Code.export$(
        Code.type(
          node.name,
          `_.Enum<${Code.quote(node.name)}, ${Code.tuple(node.getValues().map((_) => Code.quote(_.name)))} >`,
        ),
      ),
    ),
  GraphQLInputObjectType: (config, node) =>
    Code.TSDoc(
      getDocumentation(config, node),
      Code.export$(Code.interface$(node.name, renderFields(config, node))),
    ),
  GraphQLInterfaceType: (config, node) => {
    const implementors = config.typeMapByKind.GraphQLObjectType.filter(_ =>
      _.getInterfaces().filter(_ => _.name === node.name).length > 0
    )
    return Code.TSDoc(
      getDocumentation(config, node),
      Code.export$(Code.type(
        node.name,
        `_.Interface<${Code.quote(node.name)}, ${renderFields(config, node)}, ${
          Code.tuple(implementors.map(_ => `Object.${_.name}`))
        }>`,
      )),
    )
  },
  GraphQLObjectType: (config, node) =>
    Code.TSDoc(
      getDocumentation(config, node),
      Code.export$(Code.type(node.name, `_.Object<${Code.quote(node.name)}, ${renderFields(config, node)}>`)),
    ),
  GraphQLScalarType: () => ``,
  GraphQLUnionType: (config, node) =>
    Code.TSDoc(
      getDocumentation(config, node),
      Code.export$(
        Code.type(
          node.name,
          `_.Union<${Code.quote(node.name)},${
            Code.tuple(
              node
                .getTypes()
                .map(
                  (_) => dispatchToReferenceRenderer(config, _),
                ),
            )
          }>`,
        ),
      ),
    ),
})

const getDocumentation = (config: Config, node: Describable) => {
  const generalDescription = node.description
    ?? (config.TSDoc.noDocPolicy === `message` ? defaultDescription(node) : null)

  const deprecationDescription = isDeprecatableNode(node) && node.deprecationReason
    ? `@deprecated ${node.deprecationReason}`
    : null

  const enumMemberDescriptions: string[] = isEnumType(node)
    ? node
      .getValues()
      .map((_) => {
        const deprecationDescription = _.deprecationReason
          ? `(DEPRECATED: ${_.deprecationReason})`
          : null
        const generalDescription = _.description
          ? _.description
          : config.TSDoc.noDocPolicy === `message`
          ? `Missing description.`
          : null
        if (!generalDescription && !deprecationDescription) return null
        const content = [generalDescription, deprecationDescription]
          .filter((_) => _ !== null)
          .join(` `)
        return [_, content] as const
      })
      .filter((_): _ is [GraphQLEnumValue, string] => _ !== null)
      .map(([node, description]) => {
        const content = `"${node.name}" - ${description}`
        return content
      })
    : []
  const enumMemberDescription = enumMemberDescriptions.length > 0
    ? `Members\n${enumMemberDescriptions.join(`\n`)}`
    : null
  if (!enumMemberDescription && !generalDescription && !deprecationDescription) {
    return null
  }
  const content = [
    generalDescription,
    enumMemberDescription,
    deprecationDescription,
  ]
    .filter((_) => _ !== null)
    .join(`\n\n`)
  return content
}

const defaultDescription = (node: Describable) => `There is no documentation for this ${getNodeDisplayName(node)}.`

const renderFields = (config: Config, node: AnyGraphQLFieldsType): string => {
  return Code.object(Code.fields([
    ...values(node.getFields()).map((field) =>
      Code.TSDoc(
        getDocumentation(config, field),
        Code.field(field.name, renderField(config, field)),
      )
    ),
  ]))
}

const buildType = (config: Config, node: AnyClass) => {
  const { node: nodeInner, nullable } = unwrapNonNull(node)

  if (isNamedType(nodeInner)) {
    const namedTypeReference = dispatchToReferenceRenderer(config, nodeInner)
    // const namedTypeCode = `_.Named<${namedTypeReference}>`
    const namedTypeCode = namedTypeReference
    return nullable
      ? `_.Nullable<${namedTypeCode}>`
      : namedTypeCode
  }

  if (isListType(nodeInner)) {
    const fieldType = `_.List<${buildType(config, nodeInner.ofType)}>` as any as string
    return nullable
      ? `_.Nullable<${fieldType}>`
      : fieldType
  }

  throw new Error(`Unhandled type: ${String(node)}`)
}

// const getNamedType = (config: Config, node: AnyClass): GraphQLNamedType => {
//   if (isNamedType(node)) return node
//   if (isNonNullType(node)) return getNamedType(config, node.ofType)
//   if (isListType(node)) return getNamedType(config, node.ofType)
//   throw new Error(`Unhandled type: ${String(node)}`)
// }

const renderField = (config: Config, field: AnyField): string => {
  const type = buildType(config, field.type)

  const args = isGraphQLOutputField(field) && field.args.length > 0
    ? renderArgs(config, field.args)
    : null

  return `_.Field<${type}${args ? `, ${args}` : ``}>`
}

const renderArgs = (config: Config, args: readonly GraphQLArgument[]) => {
  let hasRequiredArgs = false
  const argsRendered = `_.Args<${
    Code.object(
      Code.fields(
        args.map((arg) => {
          const { node, nullable } = unwrapNonNull(arg.type)
          hasRequiredArgs = hasRequiredArgs || !nullable
          return Code.field(
            arg.name,
            nullable
              ? `_.Nullable<${dispatchToReferenceRenderer(config, node)}>`
              : dispatchToReferenceRenderer(config, node),
          )
        }),
      ),
    )
  }>`
  return argsRendered
  return Code.objectFrom({
    type: { type: argsRendered },
    // allOptional: { type: !hasRequiredArgs },
  })
}

const unwrapNonNull = (
  node: AnyClass,
): { node: AnyClass; nullable: boolean } => {
  const [nodeUnwrapped, nullable] = node instanceof GraphQLNonNull ? [node.ofType, false] : [node, true]
  return { node: nodeUnwrapped, nullable }
}

const scalarTypeMap: Record<string, 'string' | 'number' | 'boolean'> = {
  ID: `string`,
  Int: `number`,
  String: `string`,
  Float: `number`,
  Boolean: `boolean`,
}

// high level

interface Input {
  schemaModulePath?: string
  schemaSource: string
  options?: {
    TSDoc?: {
      noDocPolicy?: 'message' | 'ignore'
    }
  }
}

interface Config {
  schema: GraphQLSchema
  typeMapByKind: TypeMapByKind
  TSDoc: {
    noDocPolicy: 'message' | 'ignore'
  }
}

const resolveOptions = (input: Input): Config => {
  const schema = buildSchema(input.schemaSource)
  return {
    schema,
    typeMapByKind: getTypeMapByKind(schema),
    TSDoc: {
      noDocPolicy: input.options?.TSDoc?.noDocPolicy ?? `ignore`,
    },
  }
}

export const generateCode = (input: Input) => {
  const config = resolveOptions(input)
  const { typeMapByKind } = config

  const hasQuery = typeMapByKind.GraphQLRootTypes.find(
    (_) => _.name === `Query`,
  )
  const hasMutation = typeMapByKind.GraphQLRootTypes.find(
    (_) => _.name === `Mutation`,
  )
  const hasSubscription = typeMapByKind.GraphQLRootTypes.find(
    (_) => _.name === `Subscription`,
  )

  let code = ``

  const schemaModulePath = input.schemaModulePath ?? `graphql-client/alpha/schema`

  code += `import type * as _ from ${Code.quote(schemaModulePath)}\n\n`

  code += Code.export$(
    Code.namespace(
      `$`,
      Code.group(
        Code.export$(
          Code.interface$(
            `Index`,
            Code.objectFrom({
              Root: {
                type: Code.objectFrom({
                  Query: hasQuery ? `Root.Query` : null,
                  Mutation: hasMutation ? `Root.Mutation` : null,
                  Subscription: hasSubscription ? `Root.Subscription` : null,
                }),
              },
              objects: Code.objectFromEntries(
                typeMapByKind.GraphQLObjectType.map(_ => [_.name, Code.propertyAccess(`Object`, _.name)]),
              ),
              // unionMemberNames: Code.objectFromEntries(
              //   typeMapByKind.GraphQLUnionType.map(
              //     (_) => [_.name, Code.unionItems(_.getTypes().map(_ => Code.quote(_.name)))],
              //   ),
              // ),
              unions: {
                type: Code.objectFrom(
                  {
                    Union: {
                      type: typeMapByKind.GraphQLUnionType.length > 0
                        ? Code.unionItems(
                          typeMapByKind.GraphQLUnionType.map(
                            (_) => Code.propertyAccess(`Union`, _.name),
                          ),
                        )
                        : null,
                    },
                  },
                ),
              },
              scalars: `Scalars`,
            }),
          ),
        ),
        Code.export$(
          Code.interface$(
            `Scalars`,
            Code.objectFromEntries(typeMapByKind.GraphQLScalarType.map((_) => {
              // todo strict mode where instead of falling back to "any" we throw an error
              const type = scalarTypeMap[_.name] || `string`
              return [_.name, type]
            })),
          ),
        ),
      ),
    ),
  )

  for (const [name, types] of entries(typeMapByKind)) {
    if (name === `GraphQLScalarType`) continue

    const namespaceName = name === `GraphQLRootTypes` ? `Root` : namespaceNames[name]
    code += Code.commentSectionTitle(namespaceName)
    code += Code.export$(
      Code.namespace(
        namespaceName,
        types.length === 0
          ? `// -- no types --\n`
          : types
            .map((_) => dispatchToConcreteRenderer(config, _))
            .join(`\n\n`),
      ),
    )
  }

  return code
}

export const generateFile = async (params: {
  schemaPath: string
  typeScriptPath: string
  schemaModulePath?: string
}) => {
  // todo use @dprint/formatter
  const schemaSource = await fs.readFile(params.schemaPath, `utf8`)
  const code = generateCode({ schemaSource, ...params })
  await fs.writeFile(params.typeScriptPath, code, { encoding: `utf8` })
}
