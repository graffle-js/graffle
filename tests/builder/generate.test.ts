import { readFile } from 'fs/promises'
import { expect, test } from 'vitest'
import { generateFile } from '../../src/generator/generator.js'

test(`generates types from GraphQL SDL file`, async () => {
  await generateFile({
    schemaPath: `./tests/builder/_/schema.graphql`,
    typeScriptPath: `./tests/builder/_/schema.ts`,
  })
  expect(
    await readFile(`./tests/builder/_/schema.ts`, `utf8`),
  ).toMatchSnapshot()
})
