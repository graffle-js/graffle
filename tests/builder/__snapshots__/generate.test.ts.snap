// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates types from GraphQL SDL file 1`] = `
"export namespace $ {
export interface Index {
Root: {
Query: Root.Query
Mutation: null
Subscription: null
}
unions: {
Union: Union.FooBarUnion
}
scalars: Scalars
}
export interface Scalars {

    String: string
Int: number
Float: number
Boolean: boolean
ID: string
  
}
}

// ------------------------------------------------------------ //
//                             Root                             //
// ------------------------------------------------------------ //

export namespace Root {
export interface Query {
__typename: "Query"
string: $.Scalars["String"] | null
scalars: Object.Scalars | null
fooBarUnion: Union.FooBarUnion | null
/**
* Query enum field documentation.
*/
abcEnum: Enum.ABCEnum | null
}
}

// ------------------------------------------------------------ //
//                             Enum                             //
// ------------------------------------------------------------ //

export namespace Enum {
/**
* Enum documentation.
* 
* Members
* "A" - (DEPRECATED: Enum value A is deprecated.)
* "B" - Enum B member documentation.
* "C" - Enum C member documentation. (DEPRECATED: Enum value C is deprecated.)
*/
export type ABCEnum =
| "A"
| "B"
| "C"
}

// ------------------------------------------------------------ //
//                         InputObject                          //
// ------------------------------------------------------------ //

export namespace InputObject {
// -- no types --

}

// ------------------------------------------------------------ //
//                          Interface                           //
// ------------------------------------------------------------ //

export namespace Interface {
// -- no types --

}

// ------------------------------------------------------------ //
//                            Object                            //
// ------------------------------------------------------------ //

export namespace Object {
/**
* Object documentation.
*/
export interface Foo {
__typename: "Foo"
/**
* Field documentation.
* 
* @deprecated Field a is deprecated.
*/
a: $.Scalars["String"] | null
}

export interface Bar {
__typename: "Bar"
b: $.Scalars["Int"] | null
}

export interface Scalars {
__typename: "Scalars"
a: $.Scalars["String"] | null
b: $.Scalars["Int"] | null
c: $.Scalars["Float"] | null
d: $.Scalars["Boolean"] | null
e: $.Scalars["ID"] | null
}
}

// ------------------------------------------------------------ //
//                            Union                             //
// ------------------------------------------------------------ //

export namespace Union {
/**
* Union documentation.
*/
export type FooBarUnion =
| Object.Foo& { $$union:true}
| Object.Bar& { $$union:true}
}"
`;
